brain/detail/circuit.h

```diff
@@ -222,12 +189,335 @@ class Circuit::Impl
 
     URI getMorphologyURI(const std::string& name) const
     {
-        URI uri(_morphologySource);
+        URI uri(getMorphologySource());
         uri.setPath(uri.getPath() + "/" + name + ".h5");
         return uri;
     }
 
-    const brion::SynapseSummary& getSynapseSummary() const
+    virtual URI getMorphologySource() const = 0;
+
+    virtual const brion::URI& getCircuitSource() const = 0;
+    virtual void saveMorphologyToCache(
+        const std::string& uri, const std::string& hash,
+        neuron::MorphologyPtr morphology) const = 0;
+
+    virtual CachedMorphologies loadMorphologiesFromCache(
+        const std::set<std::string>& hashes) const = 0;
+    virtual void saveSynapsePositionsToCache(
+        const uint32_t gid, const std::string& hash,
+        const brion::SynapseMatrix& value) const = 0;
+    virtual CachedSynapses loadSynapsePositionsFromCache(
+        const Strings& keys) const = 0;
+    virtual void _findSynapsePositionsColumns() const = 0;
+    virtual const brion::SynapseSummary& getSynapseSummary() const = 0;
+    virtual const brion::Synapse& getSynapseAttributes(
+        const bool afferent) const = 0;
+    virtual const brion::Synapse& getAfferentProjectionAttributes(
+        const std::string& name) const = 0;
+    virtual const brion::Synapse* getSynapseExtra() const = 0;
+    virtual const brion::Synapse& getSynapsePositions(
+        const bool afferent) const = 0;
+    virtual const URI& getSynapseSource() const = 0;
+};
+
+class SonataCircuit : public Circuit::Impl
+{
+public:
+    explicit SonataCircuit(const URI& uri)
+        : config(uri)
+        , basePath(boost::filesystem::path(uri.getPath()).parent_path())
+    {
+        const auto& subnetworks = config.getNodes();
+        const auto& node = subnetworks.front();
+        brion::Nodes nodes(URI(basePath.string() + "/" + node.elements));
```

Comment:
This is wrong, you're assuming that paths are relative, which may not be the case.
The CircuitConfig should deal with all path substitutions and return absolute paths so client code doesn't need to deal with this stuff. As a matter of fact, I missed that you haven't dealt with relative paths at all in the CircuitConfig implementation.